File: src/vector-space-ops.js
/**
 * © 2026-2026 HeadySystems Inc. All Rights Reserved.
 * PROPRIETARY AND CONFIDENTIAL.
 */

'use strict';

/**
 * Vector math primitives for 384-dimensional embedding space.
 * All operations are pure functions; no external dependencies.
 */

const EMBEDDING_DIM = 384;

// ─── Basic operations ────────────────────────────────────────────────────────

/**
 * Dot product of two vectors.
 * @param {number[]|Float64Array} a
 * @param {number[]|Float64Array} b
 * @returns {number}
 */
function dotProduct(a, b) {
  if (a.length !== b.length) throw new RangeError(`dotProduct: dimension mismatch (${a.length} vs ${b.length})`);
  let sum = 0;
  for (let i = 0; i < a.length; i++) sum += a[i] * b[i];
  return sum;
}

/**
 * Euclidean magnitude of a vector.
 * @param {number[]|Float64Array} v
 * @returns {number}
 */
function magnitude(v) {
  let sum = 0;
  for (let i = 0; i < v.length; i++) sum += v[i] * v[i];
  return Math.sqrt(sum);
}

/**
 * Normalize a vector to unit length. Returns a new Float64Array.
 * @param {number[]|Float64Array} v
 * @returns {Float64Array}
 */
function normalize(v) {
  const mag = magnitude(v);
  const out = new Float64Array(v.length);
  if (mag === 0) return out; // zero vector stays zero
  for (let i = 0; i < v.length; i++) out[i] = v[i] / mag;
  return out;
}

/**
 * Cosine similarity between two vectors (range −1 to 1).
 * @param {number[]|Float64Array} a
 * @param {number[]|Float64Array} b
 * @returns {number}
 */
function cosineSimilarity(a, b) {
  const magA = magnitude(a);
  const magB = magnitude(b);
  if (magA === 0 || magB === 0) return 0;
  return dotProduct(a, b) / (magA * magB);
}

/**
 * Euclidean distance between two vectors.
 * @param {number[]|Float64Array} a
 * @param {number[]|Float64Array} b
 * @returns {number}
 */
function euclideanDistance(a, b) {
  if (a.length !== b.length) throw new RangeError(`euclideanDistance: dimension mismatch (${a.length} vs ${b.length})`);
  let sum = 0;
  for (let i = 0; i < a.length; i++) {
    const diff = a[i] - b[i];
    sum += diff * diff;
  }
  return Math.sqrt(sum);
}

/**
 * Element-wise addition of two vectors. Returns a new Float64Array.
 * @param {number[]|Float64Array} a
 * @param {number[]|Float64Array} b
 * @returns {Float64Array}
 */
function add(a, b) {
  if (a.length !== b.length) throw new RangeError(`add: dimension mismatch (${a.length} vs ${b.length})`);
  const out = new Float64Array(a.length);
  for (let i = 0; i < a.length; i++) out[i] = a[i] + b[i];
  return out;
}

/**
 * Element-wise subtraction (a − b). Returns a new Float64Array.
 * @param {number[]|Float64Array} a
 * @param {number[]|Float64Array} b
 * @returns {Float64Array}
 */
function subtract(a, b) {
  if (a.length !== b.length) throw new RangeError(`subtract: dimension mismatch (${a.length} vs ${b.length})`);
  const out = new Float64Array(a.length);
  for (let i = 0; i < a.length; i++) out[i] = a[i] - b[i];
  return out;
}

/**
 * Scalar multiplication. Returns a new Float64Array.
 * @param {number[]|Float64Array} v
 * @param {number} scalar
 * @returns {Float64Array}
 */
function scale(v, scalar) {
  const out = new Float64Array(v.length);
  for (let i = 0; i < v.length; i++) out[i] = v[i] * scalar;
  return out;
}

// ─── Higher-order operations ─────────────────────────────────────────────────

/**
 * Centroid (mean vector) of an array of vectors.
 * @param {Array<number[]|Float64Array>} vectors
 * @returns {Float64Array}
 */
function centroid(vectors) {
  if (!vectors || vectors.length === 0) throw new Error('centroid: empty vector array');
  const dim = vectors[0].length;
  const out = new Float64Array(dim);
  for (const v of vectors) {
    if (v.length !== dim) throw new RangeError('centroid: inconsistent dimensions');
    for (let i = 0; i < dim; i++) out[i] += v[i];
  }
  for (let i = 0; i < dim; i++) out[i] /= vectors.length;
  return out;
}

/**
 * Linear interpolation between two vectors at parameter t ∈ [0, 1].
 * @param {number[]|Float64Array} a  start
 * @param {number[]|Float64Array} b  end
 * @param {number} t  interpolation factor
 * @returns {Float64Array}
 */
function lerp(a, b, t) {
  if (a.length !== b.length) throw new RangeError(`lerp: dimension mismatch (${a.length} vs ${b.length})`);
  const out = new Float64Array(a.length);
  for (let i = 0; i < a.length; i++) out[i] = a[i] + (b[i] - a[i]) * t;
  return out;
}

/**
 * Generate a random unit vector of the given dimensionality.
 * Uses Box-Muller for Gaussian sampling to ensure uniform distribution on sphere.
 * @param {number} [dimensions=EMBEDDING_DIM]
 * @returns {Float64Array}
 */
function randomVector(dimensions = EMBEDDING_DIM) {
  const raw = new Float64Array(dimensions);
  for (let i = 0; i < dimensions; i += 2) {
    // Box-Muller transform
    const u1 = Math.random();
    const u2 = Math.random();
    const r = Math.sqrt(-2 * Math.log(u1 + Number.EPSILON));
    const theta = 2 * Math.PI * u2;
    raw[i] = r * Math.cos(theta);
    if (i + 1 < dimensions) raw[i + 1] = r * Math.sin(theta);
  }
  return normalize(raw);
}

// ─── PCA via power iteration ──────────────────────────────────────────────────

/**
 * Basic PCA using power iteration (randomised SVD-lite).
 * Projects `vectors` down to `targetDims` principal components.
 *
 * @param {Array<number[]|Float64Array>} vectors  N × D matrix (array of row vectors)
 * @param {number} targetDims  desired output dimensions
 * @returns {Float64Array[]}  projected vectors (N × targetDims)
 */
function pca(vectors, targetDims) {
  if (!vectors || vectors.length === 0) throw new Error('pca: empty input');
  const N = vectors.length;
  const D = vectors[0].length;
  if (targetDims >= D) throw new RangeError(`pca: targetDims (${targetDims}) must be < source dims (${D})`);

  // 1. Centre the data
  const mean = centroid(vectors);
  const centred = vectors.map(v => subtract(v, mean));

  // 2. Extract `targetDims` principal components via deflation + power iteration
  const components = [];
  let residual = centred.map(v => Float64Array.from(v));

  for (let k = 0; k < targetDims; k++) {
    // Random initialisation
    let pc = randomVector(D);

    // Power iteration (30 iterations is usually sufficient for embedding dims)
    for (let iter = 0; iter < 30; iter++) {
      // Compute covariance-vector product: C^T * C * pc  (= X^T (X pc))
      // Step 1: project each row onto pc → scores
      const scores = residual.map(row => dotProduct(row, pc));
      // Step 2: reconstruct direction
      const next = new Float64Array(D);
      for (let i = 0; i < N; i++) {
        for (let j = 0; j < D; j++) next[j] += scores[i] * residual[i][j];
      }
      pc = normalize(next);
    }

    components.push(pc);

    // Deflate: remove this component from residual
    residual = residual.map(row => {
      const proj = dotProduct(row, pc);
      return subtract(row, scale(pc, proj));
    });
  }

  // 3. Project original (centred) vectors onto components
  return centred.map(row => {
    const projected = new Float64Array(targetDims);
    for (let k = 0; k < targetDims; k++) projected[k] = dotProduct(row, components[k]);
    return projected;
  });
}

// ─── Exports ─────────────────────────────────────────────────────────────────

module.exports = {
  EMBEDDING_DIM,
  cosineSimilarity,
  euclideanDistance,
  dotProduct,
  normalize,
  magnitude,
  add,
  subtract,
  scale,
  centroid,
  lerp,
  randomVector,
  pca,
};

File: python/core/vector_space_ops.py
"""
Heady™ Vector Space Operations — Python SDK
© 2024-2026 HeadySystems Inc. All Rights Reserved.
PROPRIETARY AND CONFIDENTIAL.

Vector math primitives for 384-dimensional embedding space.
All operations are pure functions; no external dependencies beyond numpy.
Direct port of src/vector-space-ops.js.
"""

from __future__ import annotations

import math
import numpy as np
from typing import List, Optional, Union

# ─── Constants ────────────────────────────────────────────────────────────────

EMBEDDING_DIM: int = 384

# Type alias for vector inputs
VectorLike = Union[np.ndarray, List[float]]


def _to_array(v: VectorLike) -> np.ndarray:
    """Convert any vector-like input to a float64 numpy array."""
    if isinstance(v, np.ndarray):
        return v.astype(np.float64, copy=False)
    return np.array(v, dtype=np.float64)


# ─── Basic operations ────────────────────────────────────────────────────────


def dot_product(a: VectorLike, b: VectorLike) -> float:
    """
    Dot product of two vectors.

    Args:
        a: First vector.
        b: Second vector.

    Returns:
        Scalar dot product.

    Raises:
        ValueError: If vectors have different dimensions.
    """
    a, b = _to_array(a), _to_array(b)
    if a.shape[0] != b.shape[0]:
        raise ValueError(
            f"dot_product: dimension mismatch ({a.shape[0]} vs {b.shape[0]})"
        )
    return float(np.dot(a, b))


def magnitude(v: VectorLike) -> float:
    """
    Euclidean magnitude (L2 norm) of a vector.

    Args:
        v: Input vector.

    Returns:
        Scalar magnitude.
    """
    v = _to_array(v)
    return float(np.linalg.norm(v))


def normalize(v: VectorLike) -> np.ndarray:
    """
    Normalize a vector to unit length.

    Args:
        v: Input vector.

    Returns:
        Unit vector as float64 numpy array. Zero vector returns zero vector.
    """
    v = _to_array(v)
    mag = np.linalg.norm(v)
    if mag == 0:
        return np.zeros_like(v, dtype=np.float64)
    return v / mag


def cosine_similarity(a: VectorLike, b: VectorLike) -> float:
    """
    Cosine similarity between two vectors (range -1 to 1).

    Args:
        a: First vector.
        b: Second vector.

    Returns:
        Cosine similarity score.
    """
    a, b = _to_array(a), _to_array(b)
    mag_a = np.linalg.norm(a)
    mag_b = np.linalg.norm(b)
    if mag_a == 0 or mag_b == 0:
        return 0.0
    return float(np.dot(a, b) / (mag_a * mag_b))


def euclidean_distance(a: VectorLike, b: VectorLike) -> float:
    """
    Euclidean distance between two vectors.

    Args:
        a: First vector.
        b: Second vector.

    Returns:
        Euclidean distance.

    Raises:
        ValueError: If vectors have different dimensions.
    """
    a, b = _to_array(a), _to_array(b)
    if a.shape[0] != b.shape[0]:
        raise ValueError(
            f"euclidean_distance: dimension mismatch ({a.shape[0]} vs {b.shape[0]})"
        )
    return float(np.linalg.norm(a - b))


def add(a: VectorLike, b: VectorLike) -> np.ndarray:
    """
    Element-wise addition of two vectors.

    Args:
        a: First vector.
        b: Second vector.

    Returns:
        Sum vector as float64 numpy array.

    Raises:
        ValueError: If vectors have different dimensions.
    """
    a, b = _to_array(a), _to_array(b)
    if a.shape[0] != b.shape[0]:
        raise ValueError(
            f"add: dimension mismatch ({a.shape[0]} vs {b.shape[0]})"
        )
    return a + b


def subtract(a: VectorLike, b: VectorLike) -> np.ndarray:
    """
    Element-wise subtraction (a - b).

    Args:
        a: First vector.
        b: Second vector.

    Returns:
        Difference vector as float64 numpy array.

    Raises:
        ValueError: If vectors have different dimensions.
    """
    a, b = _to_array(a), _to_array(b)
    if a.shape[0] != b.shape[0]:
        raise ValueError(
            f"subtract: dimension mismatch ({a.shape[0]} vs {b.shape[0]})"
        )
    return a - b


def scale(v: VectorLike, scalar: float) -> np.ndarray:
    """
    Scalar multiplication.

    Args:
        v: Input vector.
        scalar: Scaling factor.

    Returns:
        Scaled vector as float64 numpy array.
    """
    return _to_array(v) * scalar


# ─── Higher-order operations ─────────────────────────────────────────────────


def centroid(vectors: List[VectorLike]) -> np.ndarray:
    """
    Centroid (mean vector) of an array of vectors.

    Args:
        vectors: List of vectors.

    Returns:
        Mean vector as float64 numpy array.

    Raises:
        ValueError: If the list is empty or dimensions are inconsistent.
    """
    if not vectors:
        raise ValueError("centroid: empty vector array")
    arrs = [_to_array(v) for v in vectors]
    dim = arrs[0].shape[0]
    for v in arrs:
        if v.shape[0] != dim:
            raise ValueError("centroid: inconsistent dimensions")
    return np.mean(arrs, axis=0).astype(np.float64)


def lerp(a: VectorLike, b: VectorLike, t: float) -> np.ndarray:
    """
    Linear interpolation between two vectors at parameter t in [0, 1].

    Args:
        a: Start vector.
        b: End vector.
        t: Interpolation factor.

    Returns:
        Interpolated vector as float64 numpy array.

    Raises:
        ValueError: If vectors have different dimensions.
    """
    a, b = _to_array(a), _to_array(b)
    if a.shape[0] != b.shape[0]:
        raise ValueError(
            f"lerp: dimension mismatch ({a.shape[0]} vs {b.shape[0]})"
        )
    return a + (b - a) * t


def random_vector(dimensions: int = EMBEDDING_DIM, rng: Optional[np.random.Generator] = None) -> np.ndarray:
    """
    Generate a random unit vector of the given dimensionality.
    Uses Gaussian sampling (Box-Muller equivalent) for uniform distribution on sphere.

    Args:
        dimensions: Number of dimensions (default: 384).
        rng: Optional numpy random generator for reproducibility.

    Returns:
        Unit vector as float64 numpy array.
    """
    if rng is None:
        rng = np.random.default_rng()
    raw = rng.standard_normal(dimensions).astype(np.float64)
    return normalize(raw)


# ─── PCA via power iteration ─────────────────────────────────────────────────


def pca(vectors: List[VectorLike], target_dims: int) -> List[np.ndarray]:
    """
    Basic PCA using power iteration (randomised SVD-lite).
    Projects vectors down to target_dims principal components.

    Args:
        vectors: N x D matrix (list of row vectors).
        target_dims: Desired output dimensions.

    Returns:
        List of projected vectors (N x target_dims).

    Raises:
        ValueError: If input is empty or target_dims >= source dims.
    """
    if not vectors:
        raise ValueError("pca: empty input")
    arrs = [_to_array(v) for v in vectors]
    N = len(arrs)
    D = arrs[0].shape[0]
    if target_dims >= D:
        raise ValueError(
            f"pca: target_dims ({target_dims}) must be < source dims ({D})"
        )

    # 1. Centre the data
    mean = centroid(arrs)
    centred = [v - mean for v in arrs]

    # 2. Extract target_dims principal components via deflation + power iteration
    components = []
    residual = [v.copy() for v in centred]

    for k in range(target_dims):
        # Random initialisation
        pc = random_vector(D)

        # Power iteration (30 iterations)
        for _ in range(30):
            # Compute covariance-vector product: X^T (X pc)
            scores = [float(np.dot(row, pc)) for row in residual]
            next_pc = np.zeros(D, dtype=np.float64)
            for i in range(N):
                next_pc += scores[i] * residual[i]
            pc = normalize(next_pc)

        components.append(pc)

        # Deflate: remove this component from residual
        for i in range(N):
            proj = float(np.dot(residual[i], pc))
            residual[i] = residual[i] - pc * proj

    # 3. Project original (centred) vectors onto components
    projected = []
    for row in centred:
        p = np.array([float(np.dot(row, comp)) for comp in components], dtype=np.float64)
        projected.append(p)

    return projected
